README.md
# node-red-contrib-events
Event broadcast and receive nodes for Node-RED.

This is an alternative take on Node-RED's `link-in` and `link-out` core nodes. Those nodes also use JavaScript events thanks to the `RED.events` object.
However, they attach/consume events named after a node ID and those links have to be manually configured.

Instead, this set of nodes name events after the input `msg.topic` and allow for wildcard links. Configuration is simply specifying the topic in the settings of the `event-in` node.

So, throw any `msg` you like to an `event-out` node. Then add `event-in` nodes set to the topic string that you want to listen for.

Then for a msg sent to any of the `event-out` nodes that has the matching `msg.topic` string, all of the `event-in` nodes will receive and output the same `msg`.

THe `event-out` node supports topic wildcards using `*`, `/` is recognised as a namespace separator as with MQTT.

For more complex requirements, there is also an `event-return` node that lets you return to the originating `event-out` that an `event-in` listened for.
This allows you to create sub-flows (sub-routines) and loops very easily.

## Notes

* For the most part, nodes like this should be avoided because then can make the tracing of your logic hard if you aren't careful.
  However, used with care, these are very powerful but simple to use nodes that can greatly simplify certain types of logic.
* The event handler is shared between the nodes. Event names are the `msg.topic` prefixed with `node-red-contrib-events/`. Return events use `node-red-contrib-events/return/<node.id>` 
* A separate event handler is needed because the Node-RED core devs want to make sure that Node-RED's own event handlers are not used by contributed nodes.
  These nodes use a separate event handler module that is shared with other nodes from me including node-red-contrib-uibuilder.
  That means that these nodes will (in the future) interact with uibuilder where needed but are not dependent.
  Other nodes could also use the same library. Whichever nodes use the library, they will all get the same event handler.
* The `/` char in the topic is used as a level (namespace) separator in the same way as with MQTT.
* Can use either glob-style (`*`, `**`) or MQTT-style (`+`, `#`) wildcards.
* The package contains some example flows. Access using the Import examples library in Node-RED.

Contact the author if you want to make use of this package's event system in other nodes.

## Installation

To install from the standard npmjs registry: `npm install @totallyinformation/node-red-contrib-events`

To install from GitHub: `npm install TotallyInformation/node-red-contrib-events`

## The nodes

### `event-out`

As the output of a flow, emits an event who's name is based on the received `msg.topic`.

There is no need for any configuration.

However, if you wish, you can:

* Provide a default topic (in case the received msg does not contain one).
* Allow the received msg to be passed through to the output.
* Allow the use of an `event-return` node so that sub-subflows (sub-routines) or loops can be created.

The node adds a `msg._eventOriginator` array property that records the `event-out` node ID that is originating the msg. This helps with debugging and tracing of data flows.
If a series of `event.out` nodes are used, each will add an ID to top of the array so that you can see the full path taken and so that the `event-return` nodes can navigate back up the tree.

The input topic is sanity checked: it must be a string and no more than 255 chars. It should not contain `*`, `#` or `+`


### `event-in`

As the trigger of a flow, listens for events from `event-out` nodes and passes on the msg received from them.

Needs to have a topic pattern specified in the settings. This can contain wildcards and so listen for multiple topics.

Wildcards can be glob or MQTT style and the `/` character denotes namespace delimiters (in the same way that MQTT does).

`*` or `+` can be used as single-level wildcards, `**` or `#` as multi-level wildcards.

### `event-return`

Allows event-based sub-flows or loops. 

Change an `event-out` node to allow output from an `event-return` node.

Then create a flow starting with an `event-in` node, do some processing and end with an `event-return` node.

The `event-return` node does not need any configuration, it uses the `msg._eventOriginator` property from the input msg
to route data back to the originating `event-out` using an internal event sender/listener based on the node ID of the `event-out` node.
In the process, it adds to top of an array `msg._eventReturner` with the ID of the return node so that you can trace backwards if needed.

You can, however, allow the node to pass messages through to an output if you want to.

You can also override the msg.topic for the output if you wish.

## Dependencies

* [@TotallyInformation/ti-common-event-handler](https://github.com/TotallyInformation/ti-common-event-handler) - Shared event handler for TotallyInformation apps. Allows wildcards for event handlers.

## Similar Nodes

* `link-in`/`link-out` are core nodes, they use Node.js's event system but are linked by source/destination node ID rather than by topic. There is currently no `return` link capability but this is [being looked into](https://discourse.nodered.org/t/link-request-node/8906/3).
* **[node-red-contrib-topic-link](https://flows.nodered.org/node/node-red-contrib-topic-link)** - uses `set`s and a map and may need a deep object copy. Does not use events. Supports MQTT-style wildcards. This is the node most similar to what I'm doing. The basic usage is similar, it is mostly the execution that is different. However, there is no `return` link capability.
* [node-red-contrib-sub-link](https://flows.nodered.org/node/node-red-contrib-sub-link) - does not use Node.js's native event system and requires configuration. Supports MQTT-style wildcards.
* [node-red-contrib-pubsub](https://flows.nodered.org/node/node-red-contrib-pubsub) - source code not published. They manually do what the native event system does for you. They manually convert msg objects to a string and back. The stringify process isn't wrapped in a try/catch which may be a little fragile. And finally, they don't support wildcard subscriptions.

## Change Log

See the [CHANGELOG](./CHANGELOG.md) file for details.

## To Do

* Allow though AND return

CHANGELOG.md
## 1.1.3 - Bug Fix (2021-09-27)

### New

* Additional example flow included for testing multi-level out and return.

### Changed

* Change return node so that topic setting doesn't override the msg.topic. The topic setting in the node will now only be used if the inbound msg has no topic.
* Previous example flows updated.

### Fixed

* [A nested loops issue](https://discourse.nodered.org/t/new-node-node-red-contrib-events-alternative-to-link-nodes/51028/35).
  
  `msg._eventOriginator` and `msg._eventReturner` are now both arrays with the latest entries at the top. This enables multi-level out and return. 
  
  Infinite loops should no longer be possible unless you deliberately mess with these properties (hint: don't!).

## 1.1.2 - Improvements & new event-return node (2021-09-26)

### New

* `event-return` node. Allows event-based sub-flows or loops. 
  
  Change an `event-out` node to allow output from an `event-return` node.

  Then create a flow starting with an `event-in` node, do some processing and end with an `event-return` node.

  The `event-return` node does not need any configuration, it uses the `msg._eventOriginator` property from the input msg
  to route data back to the originating `event-out` using an internal event sender/listener based on the node ID of the `event-out` node.

  You can, however, allow the node to pass messages through to an output if you want to.

  You can also override the msg.topic for the output if you wish.

* New example added that shows (and tests) the use of the `event-return` node.

### Changed

* Removed global variable.
* Moved the event handler to an external module [ti-common-event-handler](https://github.com/TotallyInformation/ti-common-event-handler)
  
  This module will be used by other packages including node-red-contrib-uibuilder to provide a common event handler.

## 1.0.0 - First published version (2021-09-16)

## 0.1.0 - Initial code drop


nodes/event-out.js
/** Converts an input msg to an event, passes the msg via the event.
 *  Destructured to make for easier and more consistent logic.
 * 
 * Copyright (c) 2021 Julian Knight (Totally Information)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

/** --- Type Defs ---
 * @typedef {import('../typedefs.js').runtimeRED} runtimeRED
 * @typedef {import('../typedefs.js').runtimeNodeConfig} runtimeNodeConfig
 * @typedef {import('../typedefs.js').runtimeNode} runtimeNode
 * typedef {import('../typedefs.js').myNode} myNode
 */

//#region ----- Module level variables ---- //

//const emitter = require('./libs/events.js')
const tiEvents = require('@totallyinformation/ti-common-event-handler')

/** Main (module) variables - acts as a configuration object
 *  that can easily be passed around.
 */
const mod = {
    /** @type {runtimeRED} Reference to the master RED instance */
    RED: undefined,
    /** @type {string} Custom Node Name - has to match with html file and package.json `red` section */
    nodeName: 'event-out',
    /** @type {boolean} Turn on/off node debugging */
    debug: false,
}

//#endregion ----- Module level variables ---- //

//#region ----- Module-level support functions ----- //

/** 3) Run whenever a node instance receives a new input msg
 * NOTE: `this` context is still the parent (nodeInstance).
 * See https://nodered.org/blog/2019/09/20/node-done 
 * @param {object} msg The msg object received.
 * @param {Function} send Per msg send function, node-red v1+
 * @param {Function} done Per msg finish function, node-red v1+
 */
function inputMsgHandler(msg, send, done) { // eslint-disable-line no-unused-vars
    // As a module-level named function, it will inherit `mod` and other module-level variables

    // If you need it - or just use mod.RED if you prefer:
    //const RED = mod.RED


    // Use the default topic from the settings if the msg doesn't define one
    let topic = msg.topic
    if ( (! topic) && this.topic ) topic = this.topic
    if ( ! topic ) topic = ''

    // Sanitise the topic & report errors - it must be a string and no more than 255 chars. It should not contain `*`, `#` or `+`
    let topicErrors = []
    if ( topic.length < 1 ) topicErrors.push('Topic is empty')
    if ( topic.length > 255 ) topicErrors.push('Topic is >255 characters long')
    if ( topic.includes('#') ) topicErrors.push('Topic contains a `#` wildcard character')
    if ( topic.includes('+') ) topicErrors.push('Topic contains a `+` wildcard character')
    if ( topic.includes('*') ) topicErrors.push('Topic contains a `*` wildcard character')
    if ( topicErrors.length !== 0 ) {
        done(`Input topic is invalid: ${topicErrors.join(', ')}`)
        return
    }

    // If passthrough set to 'input', send the original msg to the output port
    if ( this.passthrough === 'input' ) send(msg)

    // Ensure that the event name is unique for these nodes (since the event handler may be reused in other nodes)
    const eventName = `node-red-contrib-events/${topic}`
    
    // Add this.id to new level of _eventOriginator. Done in a way so we don't impact the through msg
    let _eo = []
    if ( msg._eventOriginator ) _eo = Array.from(msg._eventOriginator)
    _eo.unshift(this.id)

    // Emit the event
    tiEvents.emit(eventName, { ...msg, ...{_eventOriginator: _eo} })
    
    // We are done
    done()

} // ----- end of inputMsgHandler ----- //

/** 2) This is run when an actual instance of our node is committed to a flow
 * @param {runtimeNodeConfig} config The Node-RED node instance config object
 */
function nodeInstance(config) {
    // As a module-level named function, it will inherit `mod` and other module-level variables

    // If you need it - which you will here - or just use mod.RED if you prefer:
    const RED = mod.RED

    // Create the node instance - `this` can only be referenced AFTER here
    RED.nodes.createNode(this, config) 

    // Transfer config items from the Editor panel to the runtime
    this.name = config.name
    this.topic = config.topic || ''
    this.passthrough = config.passthrough || 'none'

    let s = []
    if ( this.passthrough === 'input' ) s = ['green', 'dot']
    else if ( this.passthrough === 'return' ) s = ['blue', 'ring']

    if ( mod.debug ) this.status({fill:s[0],shape:s[1],text:this.id})
    else if ( this.passthrough !== 'none' ) this.status({fill:s[0],shape:s[1],text:this.passthrough})

    // Handle v1.0 legacy
    if ( this.passthrough === false ) this.passthrough = 'none'
    else if ( this.passthrough === true ) this.passthrough = 'input'

    /** Handle incoming msg's - note that the handler fn inherits `this`
     *  The inputMsgHandler function is executed every time this instance
     *  of the node receives a msg in a flow.
     */
    this.on('input', inputMsgHandler)

    if ( this.passthrough === 'return' ) {   // if set to 'return', set up a return listener
        tiEvents.on(`node-red-contrib-events/return/${this.id}`, (msg) => {
            // Drop the top entry
            let _eo = []
            if ( msg._eventOriginator ) _eo = Array.from(msg._eventOriginator)
            _eo.shift()

            this.send({ ...msg, ...{_eventOriginator: _eo} })
        })
    }

    /** Put things here if you need to do anything when a node instance is removed
     * Or if Node-RED is shutting down.
     * Note the use of an arrow function, ensures that the function keeps the
     * same `this` context and so has access to all of the node instance properties.
     */
    this.on('close', (removed, done) => { 

        // Cancel any return listeners for this node
        tiEvents.removeAllListeners(`node-red-contrib-events/return/${this.id}`)

        done()
    })

    /** Properties of `this`
     * Methods: updateWires(wires), context(), on(event,callback), emit(event,...args), removeListener(name,listener), removeAllListeners(name), close(removed)
     *          send(msg), receive(msg), log(msg), warn(msg), error(logMessage,msg), debug(msg), trace(msg), metric(eventname, msg, metricValue), status(status)
     * Other: credentials, id, type, z, wires, x, y
     * + any props added manually from config, typically at least name and topic
     */
}

//#endregion ----- Module-level support functions ----- //

/** 1) Complete module definition for our Node. This is where things actually start.
 * @param {runtimeRED} RED The Node-RED runtime object
 */
function EventOut(RED) {
    // As a module-level named function, it will inherit `mod` and other module-level variables

    // Save a reference to the RED runtime for convenience
    mod.RED = RED

    // Register a new instance of the specified node type (2)
    RED.nodes.registerType(mod.nodeName, nodeInstance)
}

// Export the module definition (1), this is consumed by Node-RED on startup.
module.exports = EventOut

//EOF
